hipcheck/src/policy_exprs/env.rs

```diff
@@ -23,14 +32,274 @@ pub struct Env<'parent> {
 #[derive(Clone)]
 pub enum Binding {
 	/// A function.
-	Fn(Op),
+	Fn(FunctionDef),
 
 	/// A primitive value.
 	Var(Primitive),
 }
 
-/// Helper type for operation function pointer.
-type Op = fn(&Env, &[Expr]) -> Result<Expr>;
+// Ensure that type of array elements is valid with a lambda
+fn ty_check_higher_order_lambda(
+	l_ty: &FunctionType,
+	arr_ty: &ExprArrayType,
+) -> Result<ReturnableType> {
+	if let Some(arr_elt_ty) = arr_ty {
+		// Copy the lambda function type, replace ident with arr_elt_ty
+		let mut try_l_ty = l_ty.clone();
+		let first_arg = try_l_ty.arg_tys.get_mut(0).ok_or(Error::NotEnoughArgs {
+			name: "".to_owned(),
+			expected: 1,
+			given: 0,
+		})?;
+		*first_arg = Type::Primitive(*arr_elt_ty);
+		// If this returns error, means array type was incorrect for lambda
+		try_l_ty.get_return_type()
+	} else {
+		Ok(ReturnableType::Unknown)
+	}
+}
+
+fn ty_filter(args: &[Type]) -> Result<ReturnableType> {
+	let arr_ty = expect_array_at(args, 1)?;
+	let wrapped_l_ty = args.first().unwrap();
```

Comment:
Many of these type checking functions `unwrap` on indexing the array of types. It does seem like it's probably guaranteed not to panic, if the logic to check for correct number of arguments is working correctly.

But just in case that logic is broken, it would be better to return an Internal Error with an explanation of what invariant was broken, instead of panicking.